src/include/daos_pipeline.h

```diff
@@ -0,0 +1,256 @@
+
+/** types of pipeline nodes */
+#define DAOS_PIPELINE_CONDITION		0
+#define DAOS_PIPELINE_AGGREGATION	1
+
+
+/**
+ * A filter object, used to build operations for a pipeline node.
+ *
+ */
+typedef struct {
+	/**
+	 * filter type can be any of the following:
+	 *   -- functions:
+	 *      - logical functions:
+	 *          DAOS_FILTER_FUNC_EQ:		==
+	 *          DAOS_FILTER_FUNC_NE:		!=
+	 *          DAOS_FILTER_FUNC_LT:		<
+	 *          DAOS_FILTER_FUNC_LE:		<=
+	 *          DAOS_FILTER_FUNC_GE:		>=
+	 *          DAOS_FILTER_FUNC_GT:		>
+	 *          DAOS_FILTER_FUNC_LIKE:		== (reg exp.)
+	 *          DAOS_FILTER_FUNC_ISNULL:		==NULL
+	 *          DAOS_FILTER_FUNC_ISNOTNULL:		!=NULL
+	 *          DAOS_FILTER_FUNC_AND:		&&
+	 *          DAOS_FILTER_FUNC_OR:		||
+	 *      - aggeration functions:
+	 *          DAOS_FILTER_FUNC_SUM:		SUM()
+	 *          DAOS_FILTER_FUNC_MIN:		MIN()
+	 *          DAOS_FILTER_FUNC_MAX:		MAX()
+	 *          DAOS_FILTER_FUNC_AVG:		AVG()
+	 *   -- key:
+	 *          DAOS_FILTER_DKEY:	Filter object represents dkey
+	 *          DAOS_FILTER_AKEY	Filter object represents akey
+	 *   -- constant:
+	 *          DAOS_FILTER_CONST:	Filter object is a constant
+	 */
+	int		filter_type;
+	/**
+	 * Type of data. Only relevant for keys and constant filter type objects:
+	 *          DAOS_FILTER_TYPE_BINARY
+	 *          DAOS_FILTER_TYPE_STRING
+	 *          DAOS_FILTER_TYPE_INTEGER
+	 *          DAOS_FILTER_TYPE_REAL
+	 */
+	int		data_type;
+	/**
+	 * Number of parameters for this filter object. For example, for '=='
+	 * we have 2 parameters.
+	 */
+	uint32_t	num_params;
+	/**
+	 * If filtering by akey, this tells us which one.
+	 */
+	d_iov_t		akey;
+	/**
+	 * If filter object is a constant, this holds its value.
+	 */
+	d_iov_t		constant;
+	/**
+	 * If filter should only be applied starting at an offset of the data.
+	 */
+	size_t		data_offset;
+	/**
+	 * Size of the data to be filtered.
+	 */
+	size_t		data_len;
+} daos_pipeline_filter_t;
+
+/**
+ * A pipeline node, used to build a pipeline.
+ */
+typedef struct {
+	/**
+	 * Node type can be any of the following:
+	 *   -- DAOS_PIPELINE_CONDITION:
+	 *          Records in, and records (meeting condition) out
+	 *   -- DAOS_PIPELINE_AGGREGATION:
+	 *          Records in, a single value out
+	 *
+	 * NOTE: Pipeline nodes can only be chained the following way:
+	 *             (condition) --> (condition)
+	 *             (condition) --> (aggregation)
+	 */
+	int			node_type;
+	/**
+	 * Number of filters inside this pipeline node
+	 */
+	size_t			num_filters;
+	/**
+	 * Array of filters for this node
+	 */
+	daos_pipeline_filter_t	*filters;
+} daos_pipeline_node_t;
+
+/**
+ * A pipeline.
+ */
+typedef struct {
+	/**
+	 * DAOS object to which this pipeline applies
+	 */
+	daos_handle_t		oh;
+	/**
+	 * Number of nodes chained in this pipeline
+	 */
+	size_t			num_nodes;
+	/**
+	 * Array of nodes for this pipeline
+	 */
+	daos_pipeline_node_t	*nodes;
```

Comment:
If there are multiple filters here,   it is && or || ?   for example dkey1 >= 5   dkey2 >=7,  so the filter will be dkey1 >=5 && dkey2 >=7   or  dkey1 >=5 || dkey2 <=7 ?


Do we need extra field to tell this or it can only specify 1 condition for the moment?

Reply:
Multiple chained conditional filters always represent an 'and', since discarded records from a previous node can't be fetched again. Nevertheless, for now I don't envision pipelines to have more than one condition node to check data on akeys/dkeys, sinde a node can have an arbitrarily large condition. For example:

akey1 >= 5 && akey2 >=7 can be submitted in a single node with 7 filter objects:

FUNCTION_AND |  FUNCTION_GE  |  akey=akey1  |  constant=5 | FUNCTION_GE |  akey=akey2  |  constant=7

This, of course, can be represented as 2 chained nodes with 3 filter objects each:

(FUNCTION_GE  |  akey=akey1  |  constant=5)   ====>>>   (FUNCTION_GE |  akey=akey2  |  constant=7)

So, for now, I don't think there will be more than 2/3 nodes  in a pipeline.  Really just 1 or 2. Condition, and aggregation (if aggregation is needed). Nevertheless, the infrastructure of a "pipeline" can be useful for when more objects are supported in the future...

Reply:
I see. thanks. Hmm, probably add a version number to the daos_pipeline_t, in case there will be further improvements.

Reply:
im not sure what this version number will help with in case of improvements? maybe i misunderstood.
we are not storing this pipeline object anywhere so what would be the purpose?
if the pipeline API itself changes or is improved we would need to bump the DAOS API major or minor version accordingly.

Reply:
I thought this pipeline will be explained on the server side?   unless we can always make sure client and server are always in the same version.

Reply:
ah yes makes sense. 
@eberroca you can add a version number in daos_pipeline_t, so when you ship it to the server side, the server checks if it can handle such a version of the pipeline. otherwise it returns an error. this guards against server segfaults which should never happen.

Reply:
@mchaarawi Can that be done with a simple variable? Something like `uint32_t version` ? A variable that is initialized with a constant? `uint32_t version = 1`

Is this the way to go?

Reply:
I added version number to the structure. 2 variables of uint8_t type each. One for major and one for minor.